File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/tools/ClonningService.java
package org.openmole.core.implementation.tools;

import java.io.IOException;

import com.rits.cloning.Cloner;
import com.rits.cloning.IFastCloner;
import java.io.File;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.MultipleException;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.core.implementation.data.Variable;
import org.openmole.core.implementation.internal.Activator;
import org.openmole.core.model.data.IVariable;


import static org.openmole.commons.tools.io.FileUtil.*;

public class ClonningService {

    public static IVariable clone(IVariable variable) throws InternalProcessingError, UserBadDataError {
        if (variable.getValue() == null || variable.getValue().getClass().isPrimitive()) {
            return variable;
        }

        Cloner cloner = new Cloner();
        final List<Throwable> exceptions = Collections.synchronizedList(new LinkedList<Throwable>());

        cloner.registerFastCloner(File.class, new IFastCloner() {

            @Override
            public Object clone(Object o, Cloner cloner, Map<Object, Object> map) throws IllegalAccessException {
                File toClone = (File) o;
                File cloned;
                try {
                    if (toClone.isDirectory()) {
                        cloned = Activator.getWorkspace().newTmpDir();
                    } else {
                        cloned = Activator.getWorkspace().newTmpFile();
                    }

                    copy(toClone, cloned);
                    return cloned;
                } catch (IOException ex) {
                    exceptions.add(ex);
                } catch (InternalProcessingError ex) {
                    exceptions.add(ex);
                }
                return null;
            }
        });

        Object val = cloner.deepClone(variable.getValue());

        if(!exceptions.isEmpty()) {
            throw new InternalProcessingError(new MultipleException(exceptions));
        }

        return new Variable(variable.getPrototype(), val);

       /* try {

            Iterable<File> files = FileMigrator.extractFilesFromVariable(variable);

            LocalFileCache localFileCache = new LocalFileCache();

            for (File f : files) {
                File toClone = f;
                File cloned;

                if (toClone.isDirectory()) {
                    cloned = Activator.getWorkspace().newTmpDir();
                } else {
                    cloned = Activator.getWorkspace().newTmpFile();
                }

                copy(toClone, cloned);
                localFileCache.fillInLocalFileCache(f, cloned);
            }

            IVariable ret = new Variable(variable.getPrototype(), cloner.deepClone(variable.getValue()));
            FileMigrator.initFilesInVariable(ret, localFileCache);

            return ret;

        } catch (IOException ex) {
            throw new InternalProcessingError(ex);
        } catch (StackOverflowError error) {
            throw new InternalProcessingError("Unable to clonne variable " + variable.getPrototype() + " , it is to big for the library.");
        }*/
    }
}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/tools/ClonningService.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.tools

import com.rits.cloning.Cloner
import com.rits.cloning.IFastCloner
import java.io.File
import java.math.BigInteger
import java.util.logging.Level
import java.util.logging.Logger

import org.openmole.commons.tools.io.FileUtil
import org.openmole.core.implementation.data.Variable
import org.openmole.core.implementation.internal.Activator
import org.openmole.core.model.data.IVariable

object ClonningService {
    
    def clone[T](variable: IVariable[T]): IVariable[T] = {
      Logger.getLogger(ClonningService.getClass.getName).log(Level.FINE, "Clonning {0}", variable.prototype)
        
        if (variable.value == null || 
            variable.value.asInstanceOf[AnyRef].getClass.isPrimitive ||
                variable.value.asInstanceOf[AnyRef].getClass == classOf[BigDecimal] ||
                variable.value.asInstanceOf[AnyRef].getClass == classOf[BigInteger]) {
            return variable
        }
        
        val cloner = new Cloner
       // val exceptions = Collections.synchronizedList(new LinkedList<Throwable>());

        cloner.registerFastCloner(classOf[File], new IFastCloner {

            override def clone(o: Object, cloner: Cloner, map: java.util.Map[Object, Object]): Object = {
                val toClone = o.asInstanceOf[File]
                val cloned = if (toClone.isDirectory) Activator.getWorkspace.newDir else Activator.getWorkspace.newFile
                    
                FileUtil.copy(toClone, cloned)
                cloned
            }
        })

        cloner.registerImmutable(classOf[BigDecimal])
        cloner.registerImmutable(classOf[BigInteger])
        
        //val cloned = cloner.deepClone(variable.value)

       /* if(!exceptions.isEmpty()) {
            throw new InternalProcessingError(new MultipleException(exceptions));
        }*/

        new Variable(variable.prototype, cloner.deepClone(variable.value))
    }
}
